feat(job-distributor): add exp. backoff retry to `feeds.SyncNodeInfo()`

gustavogama-cll opened the pull request:
There’s a behavior that we’ve observed for some time on the NOP side where they will add/update a chain configuration of the Job Distributor panel but the change is not reflected on the service itself. This leads to inefficiencies as NOPs are unaware of this and thus need to be notified so that they may "reapply" the configuration.

After some investigation, we suspect that this is due to connectivity issues between the nodes and the job distributor instance, which causes the message with the update to be lost.

This PR attempts to solve this by adding a "retry" wrapper on top of the existing `SyncNodeInfo` method. We rely on `avast/retry-go` to implement the bulk of the retry logic. It's configured with a minimal delay of 10 seconds, maximum delay of 30 minutes and retry a total of 56 times -- which adds up to a bit more than 24 hours.

[DPA-1371](https://smartcontract-it.atlassian.net/browse/DPA-1371)

[DPA-1371]: https://smartcontract-it.atlassian.net/browse/DPA-1371?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

gustavogama-cll commented on core/config/toml/types.go:
```diff
@@ -34,10 +34,11 @@ var ErrUnsupported = errors.New("unsupported with config v2")
+	InsecureFastScrypt       *bool
+	RootDir                  *string
+	ShutdownGracePeriod      *commonconfig.Duration
+	FeedsManagerSyncInterval *commonconfig.Duration
```
New config option was added at the root level because I couldn't find a better place. Happy to move it elsewhere as per the maintainers advice.

gustavogama-cll commented on core/services/feeds/service.go:
unit tests were skipped in this draft PR as I want to get some feedback about the approach before finishing the PR

graham-chainlink commented on core/services/feeds/service.go:
```diff
@@ -1550,6 +1553,32 @@ func (s *service) isRevokable(propStatus JobProposalStatus, specStatus SpecStatu
+
+			for _, manager := range managers {
+				s.lggr.Infow("synchronizing node info", "managerID", manager.ID)
+				err := s.SyncNodeInfo(ctx, manager.ID)
```
managers/JDs can be disabled/enabled, we should avoid synching to disabled JDs, we can use the `DisabledAt` to determine it like at line 1130 of this file

graham-chainlink commented:
Hmm i wonder would this solve the connection issue?

If there is communication issue between node and JD, how would the auto sync help resolve it? It will try and it will fail right?

Alternatively would it be better to have some kind of exponential backoff retry when it does fail during the sync instead? (not that it will solve a permanent connection issue)

graham-chainlink commented on core/services/feeds/service.go:
```diff
@@ -1550,6 +1553,32 @@ func (s *service) isRevokable(propStatus JobProposalStatus, specStatus SpecStatu
 	return propStatus != JobProposalStatusDeleted && (specStatus == SpecStatusPending || specStatus == SpecStatusCancelled)
 }
 
+func (s *service) periodicallySyncNodeInfo(ctx context.Context) {
```
I think there is also an assumption in code where JD is connected to the core node when `SyncNodeInfo` is called, else it will return the error `could not fetch client` which may not be a big deal, just noise. But if we could check if the nodes are connected before synching, that would be nice.

gustavogama-cll committed:
feat(job-distributor): add exp. backoff retry to feeds.SyncNodeInfo()

There’s a behavior that we’ve observed for some time on the NOP side
where they will add/update a chain configuration of the Job
Distributor panel but the change is not reflected on the service itself.
This leads to inefficiencies as NOPs are unaware of this and thus need
to be notified so that they may "reapply" the configuration.

After some investigation, we suspect that this is due to connectivity
issues between the nodes and the job distributor instance, which causes
the message with the update to be lost.

This PR attempts to solve this by adding a "retry" wrapper on top of the
existing `SyncNodeInfo` method. We rely on `avast/retry-go` to implement
the bulk of the retry logic. It's configured with a minimal delay of
10 seconds, maximum delay of 30 minutes and retry a total of 56 times --
which adds up to a bit more than 24 hours.

Ticket Number: DPA-1371

gustavogama-cll commented:
> Alternatively would it be better to have some kind of exponential backoff retry when it does fail during the sync instead? (not that it will solve a permanent connection issue)

As discussed earlier today, I went ahead and implemented your suggestion. I ran a few manual tests and it seems to work as expected, though I had to add some extra logic around the `context` instances to get there.

I still feel the background goroutine would be more resilient. But, on the other hand, this option does not require any runtime configuration -- I think we can safely hardcode the retry parameters -- which is a _huge_ plus to me.

graham-chainlink commented:
> I still feel the background goroutine would be more resilient. But, on the other hand, this option does not require any runtime configuration -- I think we can safely hardcode the retry parameters -- which is a _huge_ plus to me.

Thanks @gustavogama-cll, yeah the background go-routine definitely has its pros, both approaches are valid, just that for me i think the retry is simpler.

graham-chainlink commented on core/services/feeds/service.go:
```diff
@@ -254,8 +258,45 @@ func (s *service) RegisterManager(ctx context.Context, params RegisterManagerPar
+
+	retryOpts := []retry.Option{
+		retry.Context(ctx),
+		retry.Delay(5 * time.Second),
+		retry.Delay(10 * time.Second),
```
Delay is configured twice.

graham-chainlink commented on core/services/feeds/service.go:
```diff
@@ -254,8 +258,45 @@ func (s *service) RegisterManager(ctx context.Context, params RegisterManagerPar
+	var ctx context.Context
+	ctx, s.syncNodeInfoCancel = context.WithCancel(context.Background())
+
+	retryOpts := []retry.Option{
```
Is there a reason we didn use `retry.BackOffDelay` ?

gustavogama-cll replied:
I understood from the docs that it's the default. But I can make it explicit if you think it's worth it.

graham-chainlink replied:
all good, i was just curious, happy either way.

graham-chainlink commented on core/services/feeds/service.go:
```diff
@@ -254,8 +258,45 @@ func (s *service) RegisterManager(ctx context.Context, params RegisterManagerPar
+		retry.Delay(5 * time.Second),
+		retry.Delay(10 * time.Second),
+		retry.MaxDelay(30 * time.Minute),
+		retry.Attempts(48 + 8), // 30m * 48 =~ 24h; plus the initial 8 shorter retries
```
Where did you derive the 8? where do we configure the shorter retries?

gustavogama-cll replied:
I did a quick back of the envelope calculation based on the exponential backoff with a factor of 2, starting with 10s: (10s + 20s + 40s ... + 8th retry) =~ 30m.

graham-chainlink replied:
hmm, where do we configure the exponential backoff? All i see here is the fixed delay per retry, so the 48 + 8 attempts have 5 or 10 seconds delay between each attempt?

gustavogama-cll replied:
What I understood (and saw in the tests) is that there's an exponential backoff that starts at `Delay` and double each time until we reach `MaxDelay`. So, we'd retry after:

1. 10s
2. 20s
3. 40s
4. 80s
5. 160s
6. 320s
7. 640s
8. 1280s
9. 1800s
10. 1800s
11. 1800s
...

> hmm, where do we configure the exponential backoff?

If you're asking about how to parameterize the delays, then I chose to leave it hardcoded. I didn't see the need to make it configurable (other than for tests; I'll need to think about this later).

graham-chainlink replied:
ah thanks! Got it, i misunderstood the library API, now i know it is exponential by default.

graham-chainlink commented on core/services/feeds/service.go:
```diff
@@ -254,8 +258,45 @@ func (s *service) RegisterManager(ctx context.Context, params RegisterManagerPar
+		}
+
+		s.syncNodeInfoCancel()
+		s.syncNodeInfoCancel = func() {}
```
Hmm wont this introduce race condition as each request that wants to update node info will try to set this variable?

gustavogama-cll replied:
Hmm.. I  think you're right. The idea behind the implementation is that there should only be one retry goroutine running but I overlooked the possibility of a race condition.

I'll think about it some more.

graham-chainlink commented on core/services/feeds/service.go:
```diff
@@ -141,6 +143,7 @@ type service struct {
 	lggr                logger.Logger
 	version             string
 	loopRegistrarConfig plugins.RegistrarConfig
+	syncNodeInfoCancel  context.CancelFunc
```
I think instead of using this to pass context, we should just have `syncNodeInfoWithRetry` accept a context as parameter, each of the caller should have a context value to pass it in.

graham-chainlink commented on core/services/feeds/service.go:
```diff
@@ -254,8 +258,45 @@ func (s *service) RegisterManager(ctx context.Context, params RegisterManagerPar
+func (s *service) syncNodeInfoWithRetry(id int64) {
+	// cancel the previous context -- and, by extension, the existing goroutine --
+	// so that we can start anew
+	s.syncNodeInfoCancel()
```
I dont think we need to do this right?

If the caller of `syncNodeInfoWithRetry` pass in their context which is scoped to a request, then we dont have to manually cancel each context. Each request should have its own retry. Eg request A should not cancel request B sync which is happening with this setup?

gustavogama-cll replied:
> Each request should have its own retry. Eg request A should not cancel request B sync which is happening with this setup?


Well, that's actually by design. The idea behind the implementation is that there should only be one retry goroutine running. The main reason I chose to go that path is to avoid overloading the JD server in case of several requests that might get queued for a retry.

Even though the retry intervals are relatively large now, I don't think it's hard to imagine a scenario where a user repeats an action a few dozen times while the JD service is offline, which could translate into dozens or even hundreds of simultaneous requests to JD. And if you've followed the profiling and optimization work we did a few weeks ago on CLO, you should know that his would not end well.

Performance reasons aside, I don't think we can use the request context for the retries. I did it before but what I noticed in those tests is that the request context is canceled -- and the associated retries aborted -- as soon as we return a response to the http client.

graham-chainlink replied:
> I don't think we can use the request context for the retries. I did it before but what I noticed in those tests is that the request context is canceled -- and the associated retries aborted -- as soon as we return a response to the http client.

ah yes good point, we cant use the request context then!

graham-chainlink replied:
> The idea behind the implementation is that there should only be one retry goroutine running.

I think this is not entirely correct?

So is it `one retry goroutine per feeds manager connection` or `one retry goroutine per whole system` , what i am seeing in this PR is the latter, i thought we want the first one ?  If that is the case, then we dont need to worry about the deadlock issue since we wont be sharing context between different feed manager connections.

gustavogama-cll replied:
You're correct, should be `one retry goroutine per feeds manager connection`. I still haven't assimilated the idea that there may be multiple feed managers. 😞 

I'll need to revisit the solution with that in mind.

gustavogama-cll committed:
review: protect cancel func access with a mutex to avoid race conditions

gustavogama-cll committed:
review: trigger retry on partial failures and support multiple job distributors

graham-chainlink commented on core/services/feeds/service.go:
```diff
@@ -402,9 +468,7 @@ func (s *service) CreateChainConfig(ctx context.Context, cfg ChainConfig) (int64
-	if err := s.SyncNodeInfo(ctx, mgr.ID); err != nil {
-		s.lggr.Infof("FMS: Unable to sync node info: %v", err)
-	}
+	s.syncNodeInfoWithRetry(mgr.ID)
```
I see, this will be a behaviour change, previously `SyncNodeInfo` is synchronous but now it is async

gustavogama-cll replied:
Yes, but `SyncNodeInfo` could fail silently -- the FeedService caller wouldn't know any better -- the only side-effect of the failure was an error message.

With the new version, the caller sees the save behavior, just faster. The clo network requests happens in the background, and we continue to log any errors.

graham-chainlink commented on core/services/feeds/service.go:
```diff
@@ -1156,6 +1214,8 @@ func (s *service) Close() error {
 		// This blocks until it finishes
 		s.connMgr.Close()
 
+		s.syncNodeInfoCancel.callAllAndClear()
```
I wonder if it is safer for this line to be called before `s.connMgr.Close()`? In case a retry interacts with a closed channel?

gustavogama-cll replied:
It shouldn't matter but here you go:
a42860e96a0d1029f5486f457bf4ec3455b1a49e

graham-chainlink reviewed (commented):
looks good

gustavogama-cll committed:
review: clear contexts before closing the connection manager
